### PR TITLE
[RV_DM]rv_dm_progbuf_read_write_execute_vseq

### PR DESCRIPTION
**Pull request by BasitMehmood**
This commit contains vseq and other realted files and verify the functioanlity of program buffer.

### CODE REVIEW

**Comment by BasitMehmood** [comment]
@jdonjdon kindly review it.

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson [hunk 1]
```diff
@@ -463,6 +463,22 @@
       stage: V1
       tests: ["rv_dm_halt_resume_whereto"]
     }
+    {
+      name: progbuf_read_write_execute
+      desc: '''
+            Verify the functionality of program buffer by writing random data on program buffer.
+
+            - Program jtag_dmi_ral.dmcontrol.haltreq to 1'b1.
+            - Acknowledge haltreq by setting tl_mem_ral.halted to 0.
```
Comment (rswarbrick) [inline hw/ip/rv_dm/data/rv_dm_testplan.hjson:472]: I know that I've said this on another PR, but I don't think this line quite makes sense. Please fix it to talk about waiting until `halted` becomes the idcode.

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson [hunk 2]
```diff
@@ -463,6 +463,22 @@
...
+            - Write random data on jtag_dmi_ral.pragram_buffer_* registers to access
+              program buffer.
```
Comment (rswarbrick) [inline hw/ip/rv_dm/data/rv_dm_testplan.hjson:474]: What does the "to access program buffer" bit mean? I'm not sure it's required.

Probably something like this would be simpler? "Fill the program buffer by writing random data to `jtag_dmi_ral.program_buffer_*` registers."

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson [hunk 3]
```diff
@@ -463,6 +463,22 @@
...
+            - send a valid abstract command(AccessRegister) by setting postexec bit set.
```
Comment (rswarbrick) [inline hw/ip/rv_dm/data/rv_dm_testplan.hjson:475]: Missing capital for "Send". Also, missing space before "(".

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson [hunk 4]
```diff
@@ -463,6 +463,22 @@
...
+            - Read tl_mem_ral.program_buf_* registers to read what data is written
+              on program buffer.
```
Comment (rswarbrick) [inline hw/ip/rv_dm/data/rv_dm_testplan.hjson:477]: This is missing the important bit! We presumably have to check that the data we wrote is what we read back.

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson [hunk 5]
```diff
@@ -463,6 +463,22 @@
...
       tests: ["rv_dm_halt_resume_whereto"]
     }
+    {
+      name: progbuf_read_write_execute
+      desc: '''
+            Verify the functionality of program buffer by writing random data on program buffer.
+
+            - Program jtag_dmi_ral.dmcontrol.haltreq to 1'b1.
+            - Acknowledge haltreq by setting tl_mem_ral.halted to 0.
+            - Write random data on jtag_dmi_ral.pragram_buffer_* registers to access
```
Comment (rswarbrick) [inline hw/ip/rv_dm/data/rv_dm_testplan.hjson:473]: Nit: I think you want "program_buffer_*" here (with an 'o')

File: hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv [hunk 6]
```diff
@@ -338,6 +338,7 @@ class rv_dm_scoreboard extends cip_base_scoreboard #(
           (!uvm_re_match("abstractcmd_*", csr.get_name())): begin
           end
           (!uvm_re_match("program_buffer_*", csr.get_name())): begin
+            do_read_check = 0;
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv:341]: Why is this needed? Since we're disabling a check, we need to add a comment that explains why.

**Review by rswarbrick — CHANGES_REQUESTED** [review]
I've added some comments about the proposed testplan change.

The code will also need quite a bit of rework, but I'd suggest you split it out into a separate PR.

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson (hunk 1 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/data/rv_dm_testplan.hjson:472]: Sorry, I was confused about this (on the other PR as well). But don't we have to see the halt signal coming out of the debug module before we tell it that we're obeying the signal? I think there's a step missing here.

File: hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv (hunk 6 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv:343]: Thanks for adding a comment. You've added:
```systemverilog
            // RO registers. Write on DM progbuf registers and read data
            // from these registers manually.
            do_read_check = 0;
```

Does that mean that we always expect the contents of `program_buffer[i]` to be the same as the thing we've seen written to `progbuf[i]`? Surely we can strengthen the checking here by teaching the predictor what the result should be?

Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv:343]: but we are checking contents of both in our sequence. Is it not enough?

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson (hunk 1 above)
Comment (BasitMehmood) [reply hw/ip/rv_dm/data/rv_dm_testplan.hjson:472]: ok i have added a step . i will do a separate PR as in this pr there are conflicts.

**Comment by BasitMehmood** [comment]
I am closing this and did a new PR of this in #21619.

**Comment by rswarbrick** [comment]
It's probably much easier to re-open this PR and force-push to the branch that it uses. There are actually merge conflicts in #21619 as well, so this should be easier.

I've just done that force-push myself (I hope it's ok!) with the following commands:
```
rjs@sole:~/work/opentitan$ git fetch basit
rjs@sole:~/work/opentitan$ git checkout -b xxx basit/progbuf_rw
Switched to a new branch 'xxx'
rjs@sole:~/work/opentitan$ git fetch origin
rjs@sole:~/work/opentitan$ # Rebase on master
rjs@sole:~/work/opentitan$ git push -f basit HEAD:progbuf_write_execute
```

File: hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv (hunk 6 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv:343]: The usual approach is to do that sort of thing in the scoreboard if possible! That way, we get checks "for free" in any other sequence that happens to use the register...

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson (hunk 1 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/data/rv_dm_testplan.hjson:505]: This looks good, thanks.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  constraint lc_hw_debug_en_c {
+    lc_hw_debug_en == lc_ctrl_pkg::On;
+  }
+  constraint scanmode_c {
+    scanmode == prim_mubi_pkg::MuBi4False;
+  }
+
+  task body();
+    Progbuf (.index(0));
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:19]: Not a big deal, but you can do this rather more simply!
```systemverilog
for (int i = 0; i < 8; i++) Progbuf(i);
```

That said, I don't really see the point in having a `Progbuf` task at all. It might be even easier to just inline its contents into the loop here.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    Progbuf (.index(5));
+    Progbuf (.index(6));
+    Progbuf (.index(7));
+  endtask
+
+  task Progbuf (input int index );
+    uvm_reg_data_t data;
+    uvm_reg_data_t r_data;
+    data = $urandom;
+      csr_wr(.ptr(jtag_dmi_ral.dmcontrol.haltreq), .value(1));
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:33]: Nit: The indentation is off from here.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      `DV_CHECK_EQ(cfg.rv_dm_vif.cb.debug_req, 1)
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:34]: It's cool that this is working, but it might be safer to wait until the `debug_req` signal becomes one. At the moment, we're also checking that the signal is raised within one cycle of the write to `haltreq`.

In general, I think you've got several vseqs that want to do the following dance:
- Write to `haltreq`
- (Possibly wait)
- Check that `debug_req` is 1
- Write to `halted`

I'd suggest you pull that out into a task. Stick it in `rv_dm_base_vseq` and use it from there.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+      csr_wr(.ptr(tl_mem_ral.halted), .value(0));
+      csr_wr(.ptr(jtag_dmi_ral.progbuf[index]), .value(data));
+      // The value in 'command' is access register command with postexec bit set.
+      csr_wr(.ptr(jtag_dmi_ral.command), .value('h002D0000));
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:38]: The magic hex value isn't lovely. Fortunately, we've got the types we need for this! I think you can essentially do the same as the code that's in `jtag_rv_debugger.sv` .

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv [hunk 11]
```diff
@@ -184,5 +184,13 @@ class rv_dm_base_vseq extends cip_base_vseq #(
...
     end
   endtask
+  // Task to halt the hart.
+  task halted();
+    uvm_reg_data_t rdata;
+    csr_wr(.ptr(jtag_dmi_ral.dmcontrol.haltreq), .value(1));
+    `DV_CHECK_EQ(cfg.rv_dm_vif.cb.debug_req, 1)
+    cfg.clk_rst_vif.wait_clks($urandom_range(0, 10));
+    csr_wr(.ptr(tl_mem_ral.halted), .value(0));
+  endtask
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv:194]: This is a good idea! To chop things into smaller pieces, maybe it makes sense to pull this out into a separate commit (and possibly PR).

Looking at the current head of master, I think you could already use this task:
- In `rv_dm_abstractcmd_status_vseq`
- In `rv_dm_base_vseq` (I doubt it would hurt to request a halt there too)
- In `rv_dm_mem_tl_access_halted_vseq`
- In `rv_dm_mem_tl_access_resuming_vseq`
- In `rv_dm_progbuf_busy_vseq`

Also (nit): missing space after the previous task.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv [hunk 12]
```diff
@@ -184,5 +184,13 @@ class rv_dm_base_vseq extends cip_base_vseq #(
       cfg.clk_rst_vif.wait_clks($urandom_range(1, 10));
     end
   endtask
+  // Task to halt the hart.
+  task halted();
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv:188]: I'd consider renaming this task. Maybe something like `request_halt`? A verb probably makes the code a bit easier to scan.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  task body();
+    uvm_reg_data_t data;
+    uvm_reg_data_t r_data;
+    data = $urandom;
+    abstract_cmd_reg_access_t cmd = '0;
+    cmd.postexec = 1;
+    cmd.transfer = 0;
+    cmd.aarpostincrement = 1;
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:24]: I haven't tested this, but... couldn't we do this in one go?
```systemverilog
abstract_cmd_reg_access_t cmd = '{postexec: 1, transfer: 0, aapostincrement: 1};
```

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    // call the task from base_vseq to halt the hart.
+    halted();
+    repeat ($urandom_range(1,10)) begin
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:27]: Is there any benefit to doing this several times? I guess that 2 might make sense, in case we're accidentally leaving state around, but I'm not sure this needs to be randomised.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+      for (int i = 0; i<8; i++) begin
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:28]: Nit: Missing spaces around `<`.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    lc_hw_debug_en == lc_ctrl_pkg::On;
+  }
+  constraint scanmode_c {
+    scanmode == prim_mubi_pkg::MuBi4False;
+  }
+
+  task body();
+    uvm_reg_data_t data;
+    uvm_reg_data_t r_data;
+    data = $urandom;
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:20]: Don't we want to randomise this to different values for the different entries in progbuf? I think this probably needs to go in the `for` loop.

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv (hunk 9 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv]: Thanks for pulling it out into a task. It might make sense to open a separate PR (in parallel) that adds this task and uses it in the places we're already doing the work.

We can always rebase whichever PR gets merged second :-)

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv (hunk 11 above)
Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv]: Ok. I will made separate PR to add request_halt() task in rv_dm_base_vseq. but let's not use this task in other vseqs that are already merged yet. First let's see if it is ok to add this task in base_vseq then after that we can use this task in other vseqs in another PR. Is it fine?

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv (hunk 13 above)
Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:23]: It is not working in this way.

**Comment by BasitMehmood** [comment]
I don't know why (Run DV smoke test) is failing?

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv (hunk 11 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv]: Oh sorry! I saw the PR (#21663) before I read this. I'd suggest using the task immediately. I can't really see a downside, and it should hopefully let us get rid of some duplicated code.

For *this* PR, we do end up with a bit of a race condition. I would probably suggest amending this PR so that it contains the commit from #21663 and then the commit that we've currently got (amended to use the new `request_halt` task). Add a note in the description of this PR mentioning that it depends on #21663 and that any review should only look at the last commit. Then reviewers (me!) will end up only looking at each line of code in one PR, and everything will get merged in a sensible order.

**Comment by rswarbrick** [comment]
The DV smoke test is failing because the code doesn't compile. I think you're using a new `request_halt` task without actually declaring it :-)

The simple solution is probably to do the 2-commit approach that I [suggested above](https://github.com/lowRISC/opentitan/pull/21544#discussion_r1501692117).

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv (hunk 11 above)
Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv]: Ok , I did two commits. In first one I added request_halt task in base_vseq as in #21663 and use it in other sequences as you suggested. Then in 2nd commit I added in description that "in last commit we added a task in base_vseq and we are using that task in our progbuf_read_write_execute_vseq". Is it ok?

Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv]: Sounds like a good idea, but this PR currently only has 1 commit... :-)

The easiest approach is probably to rebase your `progbuf_write_execute` branch on top of your `halt_task` branch and then re-push. Once all the merge conflicts are sorted out, you end up with a pair of commits that can be comfortably merged.

Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv]: Oh my bad. Actually i did PR of "progbuf_write_execute" branch from one local machine and the PR of request_halt branch from synopsys server. because we got access of git on synopsys later. So when we got the access, I started doing PR from directly synopsys instead of local server. So now there are two branches but from different servers.

Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv]: Ha, that sort of thing is confusing! On the plus side, both servers can see your remote. So something like `git fetch upstream; git rebase upstream/halt_task` might work. (You'll probably need to tweak the remote name)

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv (hunk 13 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:23]: Hmm, I've just tried, and the following line compiles happily:
```systemverilog
    abstract_cmd_reg_access_t cmd = '{zero1: 0, aarsize: 0, aarpostincrement: 1'b1,
                                      postexec: 1'b1, transfer: 0, write: 0, regno: 0};
```

But there's some work to do before we use it! Can you write down a 1-sentence comment that explains exactly what we're doing? What I can see at the moment is:

- Write `data` to `progbuf[i]`
- Write an unexplained command to `command` with `postexec=1`
- Read from `program_buffer[i]`
- Check that we just saw `data`.

I think you'll need to explain what command we're trying to run. I suspect it doesn't really matter? But it would probably be good to say this explicitly.

And what is the side effect of running the abstract command? At a quick glance, it looks like the DMI write (`csr_wr(.ptr(jtag_dmi_ral.progbuf[i]), .value(data));`) and the TL-UL read (`csr_rd(.ptr(tl_mem_ral.program_buffer[i]), .value(r_data));`) are actually backed by the same data! Was the command needed at all?

Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:23]: Oh, sorry. There's also a data type problem. I think that `abstract_cmd_reg_access_t` gives values for just *some* of the bits that we want to write. To make it work correctly, I think you'd need write something like what's currently in `jtag_rv_debugger.sv` (e.g. the `abstract_cmd_reg_read` task).

I strongly suspect this is further evidence that the contents of the command you're writing didn't matter anyway :-D

Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:23]: Oops, you are right. I am sorry. I read in spec that to execute program buffer we need to give command with postexec bit set and transfer bit unset. So that is why i was giving that particular command but when i checked with just abstract command without having postexec and transfer bit set, it is working. Thanks

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv (hunk 11 above)
Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_base_vseq.sv]: @rswarbrick i think i messed up things in doing that.

**Comment by rswarbrick** [comment]
Hmm, something has gone a little bit awry on the git side. I've just pulled the head of this PR and rebased onto `master` and will push the result. We'll then only have 2 commits, rather than 51(!)

**Comment by rswarbrick** [comment]
Something has got a little bit shuffled around. I think what we want is probably two commits:
1. Define `request_halt` and use it in places that might need it. This should come from #21663, (and that PR needs a bit of extending)
2. Define `progbuf_read_write_execute_vseq`

I think there are probably some other improvements that we can make, but it looks to me like the best thing is to sort out the basic structure first!

**Comment by BasitMehmood** [comment]
> Something has got a little bit shuffled around. I think what we want is probably two commits:
> 
> 1. Define `request_halt` and use it in places that might need it. This should come from [[RV_DM]rv_dm_base_vseq #21663](https://github.com/lowRISC/opentitan/pull/21663), (and that PR needs a bit of extending)
> 2. Define `progbuf_read_write_execute_vseq`
> 
> I think there are probably some other improvements that we can make, but it looks to me like the best thing is to sort out the basic structure first!

 I have defined halt_task in #21663 and use it in vseqs that need it, as you said. But for this PR 'progbuf_write_execute' there are  changes in vseqs that use that task. So my concern is how to delete those files of vseqs in which I use that halt_task? 
 Because in this PR we only want 'progbuf_read_write_execute_vseq' file.

**Comment by BasitMehmood** [comment]
@rswarbrick we can first merge #21663 branch where we defined halt_task and use it in vseqs. Then we can remove extra files(vseqs) in this branch that also uses that task, we can merge this PR as well. what do you think?

**Comment by rswarbrick** [comment]
All sorted, thanks. Let's rebase onto the current head of master and I think we should be good to review and merge :-)

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+  task body();
+    uvm_reg_data_t data;
+    uvm_reg_data_t r_data;
+    // Call the task from base_vseq to halt the hart.
+    request_halt();
+    repeat (2) begin
+      for (int i = 0; i < 8; i++) begin
+        data = $urandom;
+        csr_wr(.ptr(jtag_dmi_ral.progbuf[i]), .value(data));
+        csr_wr(.ptr(jtag_dmi_ral.command), .value(data));
```
Comment (rswarbrick) [inline hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:26]: Here we're writing a random value into the `command` register. That seems surprising to me! Do we need to send a command at all?

File: hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv (hunk 6 above)
Comment (rswarbrick) [reply hw/ip/rv_dm/dv/env/rv_dm_scoreboard.sv:347]: Looking back to this, I still think that it's reasonable for the scoreboard to have a prediction about what we see on a read of the program buffer (we expect the stuff that we can find in the program buffer when we use a backdoor).

I think this is still open.

**Review by rswarbrick — COMMENTED** [review]
I'm still a little unconvinced

File: hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv (hunk 17 above)
Comment (BasitMehmood) [reply hw/ip/rv_dm/dv/env/seq_lib/rv_dm_progbuf_read_write_execute_vseq.sv:26]: Yes, we can skip this command.

**Comment by BasitMehmood** [comment]
There is a check in scoreboard line 363. Which compare the value that is written, against the csr.get_mirrored_value(). But this check failed because item.d_data has written value but in csr.get_mirrored_value() has 0 value. I think this check is for all tl_mem_ral csrs and compare against the written values.

**Comment by rswarbrick** [comment]
This is unchecked, but can't we simplify this a bit? I'm imagining something like
```systemverilog
  task progbuf_rw(int idx);
    uvm_reg_data_t data = $urandom;
    uvm_reg_data_t r_data;

    csr_wr(.ptr(jtag_dmi_ral.progbuf[i]), .value(data));
    csr_rd(.ptr(tl_mem_ral.program_buffer[i]), .value(r_data));

    `DV_CHECK_EQ(data, r_data)
  endtask

  task body();
    // Call the task from base_vseq to halt the hart.
    request_halt();
    for (int i = 0; i < 8; i++) progbuf_rw(i);
  endtask
```

This is checking that DMI writes become visible over TileLink. Do we expect it to work in the other direction as well? If so, maybe we should be randomising the direction?

**Comment by BasitMehmood** [comment]
> This is unchecked, but can't we simplify this a bit? I'm imagining something like
> 
> ```systemverilog
>   task progbuf_rw(int idx);
>     uvm_reg_data_t data = $urandom;
>     uvm_reg_data_t r_data;
> 
>     csr_wr(.ptr(jtag_dmi_ral.progbuf[i]), .value(data));
>     csr_rd(.ptr(tl_mem_ral.program_buffer[i]), .value(r_data));
> 
>     `DV_CHECK_EQ(data, r_data)
>   endtask
> 
>   task body();
>     // Call the task from base_vseq to halt the hart.
>     request_halt();
>     for (int i = 0; i < 8; i++) progbuf_rw(i);
>   endtask
> ```
> 
> This is checking that DMI writes become visible over TileLink. Do we expect it to work in the other direction as well? If so, maybe we should be randomising the direction?

No, we are just checking that writes on dmi should visible on tilelink.

**Comment by BasitMehmood** [comment]
> I agree that this you're correctly describing what the code currently does :-) But I'm trying to say that this is currently testing a reasonably trivial property ("writes to X are reflected in Y"). For very little effort, I think we could test a stricter version (adding "... and vice versa"). Unless we don't believe that property should hold, I propose that we test it.

We can only write to DMI and can read from TL. As tl_mem_ral registers are read-only.  So, we can't do vice versa.

**Comment by BasitMehmood** [comment]
@rswarbrick now what's your thoughts on this PR?

**Comment by BasitMehmood** [comment]
> Sorry for the slow reply.
> 
> But I'm clearly missing something! Are you saying that it's impossible to do a CSR read over the DMI interface? I don't _think_ that's true. Note that `jtag_dmi_reg_frontdoor.sv` believes that it's possible to do such a read.

Oh there was misunderstanding. I thought your 'vice versa' means that writing on tilelink and reading from DMI. So that is why i said that we can't do that because tl_mem_ral registers are read only. But if you are trying to say that write on DMI and read from tilelink and from DMI as well?  If so, that is possible. We can read from DMI interface as well.

**Comment by BasitMehmood** [comment]
@rswarbrick i have make changes in vseq as you said to add read csr over DMI interface and randomize the read direction. Also randomize the both directions. Make changes in testpoint entry according to implementation that we have done.

**Comment by rswarbrick** [comment]
I'm so sorry: I wrote something rather silly yesterday. You were pointing out that the TL CSR interface here is read-only, so it only makes sense to send data in one direction. And you're completely right!

Your recent change is completely sensible: trying to interpret my confused message in a way that makes sense! But I was just being stupid... The best thing is probably to drop the "else" case in your if/else block.

Otherwise, the change looks completely sensible. I'm really sorry for wasting your time.

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson [hunk 18]
```diff
@@ -505,12 +505,11 @@
             - Acknowledge haltreq by setting tl_mem_ral.halted to 0.
             - Fill the program buffer by writing random data to jtag_dmi_ral.progbuf0-7
               registers.
-            - Send a valid abstract command (AccessRegister) by setting postexec bit.
```
[the remark this answers is not in the document]
Comment (BasitMehmood) [reply hw/ip/rv_dm/data/rv_dm_testplan.hjson:508]: > I didn't notice this change when reviewing a second ago. I think the testpoint was added by you in the first place. Was there a reason that you initially suggested we should send an abstract command?
> 
> I don't have strong feelings about whether it should be there, but want to make sure we don't drop a step by accident.

Yes we do not need to send abstract command for writing something and reading back. We need to send abstract command if we want to execute program buffer.

**Comment by BasitMehmood** [comment]
> I'm so sorry: I wrote something rather silly yesterday. You were pointing out that the TL CSR interface here is read-only, so it only makes sense to send data in one direction. And you're completely right!
> 
> Your recent change is completely sensible: trying to interpret my confused message in a way that makes sense! But I was just being stupid... The best thing is probably to drop the "else" case in your if/else block.
> 
> Otherwise, the change looks completely sensible. I'm really sorry for wasting your time.

I did commit now to drop changes that i committed earlier. Kindly review it.

File: hw/ip/rv_dm/data/rv_dm_testplan.hjson (hunk 18 above)
[the remark this answers is not in the document]
Comment (rswarbrick) [reply hw/ip/rv_dm/data/rv_dm_testplan.hjson:508]: Ok, thanks for the explanation. I think I understand properly now.

**Review by rswarbrick — APPROVED** [review]
